File: src/main/java/main/commands/prefixcommands/audio/JoinCommand.java
package main.commands.prefixcommands.audio;

import main.audiohandlers.RobertifyAudioManager;
import main.commands.prefixcommands.CommandContext;
import main.commands.prefixcommands.ICommand;
import main.constants.Toggles;
import main.utils.RobertifyEmbedUtils;
import main.utils.json.restrictedchannels.RestrictedChannelsConfig;
import main.utils.json.toggles.TogglesConfig;
import main.utils.locale.LocaleManager;
import main.utils.locale.RobertifyLocaleMessage;
import net.dv8tion.jda.api.entities.*;
import net.dv8tion.jda.api.entities.channel.concrete.TextChannel;
import net.dv8tion.jda.api.exceptions.InsufficientPermissionException;
import net.dv8tion.jda.internal.utils.tuple.Pair;

import javax.script.ScriptException;

public class JoinCommand implements ICommand {
    @Override
    public void handle(CommandContext ctx) throws ScriptException {
        final Message msg = ctx.getMessage();
        final Member member = ctx.getMember();
        final Guild guild = ctx.getGuild();
        final GuildVoiceState memberVoiceState = member.getVoiceState();
        final GuildVoiceState selfVoiceState = guild.getSelfMember().getVoiceState();

        msg.replyEmbeds(handleJoin(guild, ctx.getChannel(), memberVoiceState, selfVoiceState))
                .queue();
    }

    public MessageEmbed handleJoin(Guild guild, TextChannel textChannel, GuildVoiceState memberVoiceState, GuildVoiceState selfVoiceState) {
        if (!memberVoiceState.inAudioChannel()) {
            return RobertifyEmbedUtils.embedMessage(guild, RobertifyLocaleMessage.GeneralMessages.USER_VOICE_CHANNEL_NEEDED)
                    .build();
        }

        final var channel = memberVoiceState.getChannel();

        if (new TogglesConfig(guild).getToggle(Toggles.RESTRICTED_VOICE_CHANNELS)) {
            final var restrictedChannelsConfig = new RestrictedChannelsConfig(guild);
            final var localeManager = LocaleManager.getLocaleManager(guild);
            if (!restrictedChannelsConfig.isRestrictedChannel(channel.getIdLong(), RestrictedChannelsConfig.ChannelType.VOICE_CHANNEL)) {
                return RobertifyEmbedUtils.embedMessage(guild, localeManager.getMessage(RobertifyLocaleMessage.GeneralMessages.CANT_JOIN_CHANNEL) +
                        (!restrictedChannelsConfig.getRestrictedChannels(
                                RestrictedChannelsConfig.ChannelType.VOICE_CHANNEL
                        ).isEmpty()
                                ?
                                localeManager.getMessage(RobertifyLocaleMessage.GeneralMessages.RESTRICTED_TO_JOIN, Pair.of("{channels}", restrictedChannelsConfig.restrictedChannelsToString(
                                        RestrictedChannelsConfig.ChannelType.VOICE_CHANNEL
                                )))
                                :
                                localeManager.getMessage(RobertifyLocaleMessage.GeneralMessages.NO_VOICE_CHANNEL)
                        )
                ).build();
            }
        }

        final var musicManager = RobertifyAudioManager.getInstance().getMusicManager(guild);
        final var placeholderPair = Pair.of("{channel}", channel.getAsMention());

        if (memberVoiceState.getChannel() == selfVoiceState.getChannel())
            return RobertifyEmbedUtils.embedMessage(guild, RobertifyLocaleMessage.JoinMessages.ALREADY_JOINED, placeholderPair).build();
        try {
            RobertifyAudioManager.getInstance()
                    .joinAudioChannel(channel, musicManager);
        } catch (IllegalStateException e) {
            return RobertifyEmbedUtils.embedMessage(guild, RobertifyLocaleMessage.JoinMessages.CANT_JOIN, placeholderPair)
                    .build();
        } catch (InsufficientPermissionException e) {
            return RobertifyEmbedUtils.embedMessage(channel.getGuild(), RobertifyLocaleMessage.GeneralMessages.INSUFFICIENT_PERMS_TO_JOIN, placeholderPair).build();
        }
        return RobertifyEmbedUtils.embedMessage(guild, RobertifyLocaleMessage.JoinMessages.JOINED, placeholderPair).build();
    }

    @Override
    public String getName() {
        return "join";
    }

    @Override
    public String getHelp(String prefix) {
        return """
                Use this command to forcefully move the bot into your voice channel.

                *NOTE: This command can be made DJ only by using* `toggles dj join`""";
    }
}

File: src/main/kotlin/main/commands/slashcommands/audio/JoinCommand.kt
package main.commands.slashcommands.audio

import dev.minn.jda.ktx.util.SLF4J
import main.audiohandlers.RobertifyAudioManager
import main.utils.RobertifyEmbedUtils.Companion.editEmbed
import main.utils.RobertifyEmbedUtils.Companion.replyEmbed
import main.utils.component.interactions.slashcommand.AbstractSlashCommand
import main.utils.component.interactions.slashcommand.models.SlashCommand
import main.utils.locale.messages.GeneralMessages
import main.utils.locale.messages.JoinMessages
import net.dv8tion.jda.api.events.interaction.command.SlashCommandInteractionEvent

class JoinCommand : AbstractSlashCommand(
    SlashCommand(
        name = "join",
        description = "Force the bot to join the audio channel you're currently in"
    )
) {

    companion object {
        private val logger by SLF4J
    }

    override fun handle(event: SlashCommandInteractionEvent) {
        handleJoin(event)
    }

    private fun handleJoin(
        event: SlashCommandInteractionEvent
    ) {
        val guild = event.guild!!
        val memberVoiceState = event.member!!.voiceState!!
        val selfVoiceState = guild.selfMember.voiceState!!

        if (!memberVoiceState.inAudioChannel())
            return event.replyEmbed(GeneralMessages.USER_VOICE_CHANNEL_NEEDED)
                .setEphemeral(true)
                .queue()

        val channel = memberVoiceState.channel!!
        val musicManager = RobertifyAudioManager[guild]
        val placeholderPair = Pair("{channel}", channel.asMention)

        if (selfVoiceState.inAudioChannel() && memberVoiceState.channel!!.id == selfVoiceState.channel!!.id)
            return event.replyEmbed(
                JoinMessages.ALREADY_JOINED,
                placeholderPair
            ).setEphemeral(true)
                .queue()

        event.replyEmbed(JoinMessages.ATTEMPTING_TO_JOIN, placeholderPair).queue { message ->
            if (RobertifyAudioManager.joinAudioChannel(channel, musicManager, hookMessage = message)) {
                message.editEmbed(guild, JoinMessages.JOINED, placeholderPair).queue()
            }
        }
    }

    override val help: String
        get() = """
                Use this command to forcefully move the bot into your voice channel.

                *NOTE: This command can be made DJ only by using* `toggles dj join`"""
}